### ISSUE TITLE
Potential bug in timer due date checker scheduling

### ISSUE THREAD

**Issue by mustafadagher** [camunda/zeebe#17227]
**Summary**

> Try to answer the following as best as possible

- How often does the test fail? [once](https://github.com/camunda/zeebe/actions/runs/8520495239/attempts/1)
- Does it block your work? no
- Do we suspect that it is a real failure? not sure.

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
 shouldTriggerOnlyTimerStartEvent (Flaky Test): io.camunda.zeebe.engine.processing.timer.TimerStartEventTest
java.lang.AssertionError: 

Expecting ArrayList:
  []
to contain only:
  ["timer_start"]
but could not find the following element(s):
  ["timer_start"]

	at io.camunda.zeebe.engine.processing.timer.TimerStartEventTest.shouldTriggerOnlyTimerStartEvent(TimerStartEventTest.java:883)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted
No clue!
**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
https://productionresultssa2.blob.core.windows.net/actions-results/0ae2319b-b6c7-4386-ad29-60d7cae419b8/workflow-job-run-b8245a8d-c8fe-58cd-5361-3fdfc2c073c3/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-04-02T09%3A54%3A54Z&sig=PJkLcZ0vN2NDFsF46Y5K%2FrvoQJI8KncbGDktLQJPwtA%3D&sp=r&spr=https&sr=b&st=2024-04-02T09%3A44%3A49Z&sv=2021-12-02
</pre>
</details>

**Comment by oleschoenburg** [comment]
This may be an actual bug :disappointed: 

Possible explanation:
1. Start with an empty state
2. A timer is added. Scheduling sees a a planned execution that is close by and does not schedule something new.
3. Planned execution runs but does not find newly added timer (e.g. because it's not committed yet or commit isn't observed).
4. As a result, nothing is scheduled and the newly added timer is missed.

**Comment by korthout** [comment]
That reads as if this case can also occur on non-empty state

**Comment by korthout** [comment]
@oleschoenburg and I discussed the issue a bit more, and it's unclear whether the bug occurs as described above (which would mean that the execute method cannot read all committed data) or that the issue is slightly different. 

However, it's unlikely that this is a flaky test. This would require the system to have a full 5 second pause.

**Comment by oleschoenburg** [comment]
I think the initial scenario is wrong. Instead, I think this is what happens:
1. The initial execution is scheduled
2. The initial execution occurs immediately, before the `nextExecution` is set.
3. The initial execution finishes, not scheduling something new (nothing in the state yet)
4. The scheduling of the initial execution finishes, seeing that `nextExecution` is still null and sets it

Afterwards, `nextExecution` holds the initially scheduled execution and subsequent scheduling will always decide that new scheduling is not necessary.
This is a race condition between scheduling and execution of the initial execution only.

**Comment by oleschoenburg** [comment]
And here are logs that show this exact sequence:
```
19:41:28.104 [AsyncProcessingScheduleActor-1] DEBUG io.camunda.zeebe.engine.DueDateChecker{timer} - Executing now, forgetting execution null
19:41:28.106 [StreamProcessor-1] DEBUG io.camunda.zeebe.engine.DueDateChecker{timer} - Scheduled new execution NextExecution[scheduledFor=1712079688096, task=io.camunda.zeebe.stream.impl.ExtendedProcessingScheduleServiceImpl$AsyncScheduledTask@43ba5c37] as a replacment for null
19:41:28.107 [AsyncProcessingScheduleActor-1] DEBUG io.camunda.zeebe.stream - Scheduled task with delay PT1S: io.camunda.zeebe.stream.impl.ProcessingScheduleServiceImpl$$Lambda/0x00007f44c8592ec8@72739e5a
19:41:28.439 [StreamProcessor-1] DEBUG io.camunda.zeebe.engine.DueDateChecker{timer} - Not scheduling a new execution for 1712079689414, next is closed enough: NextExecution[scheduledFor=1712079688096, task=io.camunda.zeebe.stream.impl.ExtendedProcessingScheduleServiceImpl$AsyncScheduledTask@43ba5c37]
```
